src/rtree.c

```diff
@@ -66,27 +67,37 @@ rtree_delete_subtree(tsdn_t *tsdn, rtree_t *rtree, rtree_elm_t *node,
 
 void
 rtree_delete(tsdn_t *tsdn, rtree_t *rtree) {
-	if (rtree->root_pun != NULL) {
-		rtree_delete_subtree(tsdn, rtree, rtree->root, 0);
+	rtree_elm_t *rtree_root = (rtree_elm_t *)atomic_load_p(&rtree->root,
+	    ATOMIC_RELAXED);
+	if (rtree_root != NULL) {
+		rtree_delete_subtree(tsdn, rtree, rtree_root, 0);
 	}
 }
 #endif
 
 static rtree_elm_t *
 rtree_node_init(tsdn_t *tsdn, rtree_t *rtree, unsigned level,
-    rtree_elm_t **elmp) {
+    atomic_p_t *elmp) {
 	rtree_elm_t *node;
 
 	malloc_mutex_lock(tsdn, &rtree->init_lock);
-	node = atomic_read_p((void**)elmp);
+	/*
+	 * If *elmp is non-null, then it was initialized with the init lock
+	 * held, so we can get by with `relaxed` here.
+	 */
+	node = atomic_load_p(elmp, ATOMIC_RELAXED);
 	if (node == NULL) {
 		node = rtree_node_alloc(tsdn, rtree, ZU(1) <<
 		    rtree_levels[level].bits);
 		if (node == NULL) {
 			malloc_mutex_unlock(tsdn, &rtree->init_lock);
 			return NULL;
 		}
-		atomic_write_p((void **)elmp, node);
+		/*
+		 * Even though we hold the lock, a later reader might not; we
+		 * need release semantics.
+		 */
+		atomic_store_p(elmp, node, ATOMIC_RELEASE);
```

Comment:
Since we're releasing the mutex right after this, do I understand that the release semantics here would be unnecessary on total-store-order machines, but here you're guaranteeing correct operation on partial-store-order machines?